modules/core/src/edu/kit/iti/algover/proof/ProofNode.java

```diff
@@ -131,171 +105,69 @@ public SymbolTable getAllSymbols() {
-     *
-     ****************************************************************************************************************/
-
-    public VariableAssignment getVariableAssignments() {
-        return variableAssignments;
-    }
-
-    public void setVariableAssignments(VariableAssignment variableAssignments) {
-        this.variableAssignments = variableAssignments;
+    public Sequent getSequent() {
+        return sequent;
     }
 
-    /**
-     * @param varname
-     * @return value of variable if it exists
-     */
-    public Value getVariableValue(Variable varname) {
-        return variableAssignments.getValue(varname);
-
+    public ProofNode getParent() {
+        return parent;
     }
 
-    /**
-     * Lookup the type of the variable in the type map
-     *
-     * @param varname
-     * @return
-     */
-    public Type getVariableType(Variable varname) {
-        Type t = this.getAssignments().getType(varname);
-        if (t == null) {
-            throw new RuntimeException("Variable " + varname + " must be declared first");
-        } else {
-
-            return t;
-        }
+    public PVC getPVC() {
+        return pvc;
     }
 
-    public VariableAssignment getAssignments() {
-        return this.variableAssignments;
+    public ProofRuleApplication getProofRuleApplication() {
+        return ruleApplication;
     }
 
     /**
-     * Add a variable declaration to the type map (TODO Default value in map?)
+     * Get the list of all children of this proof node.
      *
-     * @param name
-     * @param type
-     * @throws NullPointerException
-     */
-    public void declareVariable(Variable name, Type type) {
-        this.getAssignments().declare(name, type);
-    }
-
-    /**
-     * Set the value of a variable in the value map
+     * This returns null during script interpretation as the proof tree grows
      *
-     * @param name
-     * @param value
+     * @return an immutable view to the list of children, null if not yet fully expanded
      */
-    public void setVariableValue(Variable name, Value value) {
-        getAssignments().assign(name, value);
-    }
-
-    /**
-     * Enter new variable scope and push onto stack
-     */
-    public VariableAssignment enterScope() {
-        variableAssignments = variableAssignments.push();
-        return variableAssignments;
-    }
-
-
-    public VariableAssignment exitScope() {
-        this.variableAssignments = this.variableAssignments.pop();
-        return variableAssignments;
-    }
-
-    public VariableAssignment enterScope(VariableAssignment va) {
-        variableAssignments = variableAssignments.push(va);
-        return variableAssignments;
-    }
-
-    /**
-     * non-obligatory label for handling cases
-     */
-    public String getLabel() {
-        return label;
+    public List<ProofNode> getChildren() {
+        return children == null ? null : Collections.unmodifiableList(children);
```

Comment:
this is the line i mentioned in my comment. do we rly want to return null here?

Reply:
Yes

`getChildren() == null` means no rule applied or error during rule application

`getChildren().isEmpty()` means that a closing rule has been successfully applied.

Of course, we could use a different kind of mechanism ...

Reply:
Moreover, I discovered that I implemented an alternative method `getSuccessors()` never returning `null` but the empty list instead.